### ISSUE TITLE
Double-to-decimal in ICU4X

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#166]
In ICU4C, we use a [short, fast algorithm](https://github.com/unicode-org/icu/blob/release-67-1/icu4c/source/i18n/number_decimalquantity.cpp#L433) when we need 14 or fewer significant digits from a double, and [google/double-conversion](https://github.com/google/double-conversion), which implements Grisu3, if we need more than 14 significant digits.

@zbraniecki pointed to https://github.com/dtolnay/ryu, a Rust library that implements a 2018 algorithm, Ryū, that outperforms Grisu3 and the Rust standard library `to_string`.  I also expect that Ryū is probably smaller in code size than the standard library, but I have not tested this.

First question: does ICU4X depend on Ryū for double-to-decimal (feature-gated), or do we make ICU4X implement FromStr, and let the user pick how they want to do the conversion themselves?

- Pro: Directly depending on Ryū might make the library easier to use and more consistent in Rust
- Con: Double-to-decimal isn't a core competency of ICU4X, so maybe we should not include it

Second question: how does this affect FFI or WebAssembly?  For example, in JavaScript, maybe we want to use Number.prototype.toString so that we don't need to ship *any* double-to-decimal code in WASM.  On the other hand, in C++, where there is no standard library function that does this\*, maybe users would want ICU4X to handle this problem on our end.

Third question: do we implement ICU4C's fastpath algorithm in ICU4X?

*\* There is sprintf, but this requires you to know a fixed number of decimal points. A general double-to-decimal algorithm akin to Number.prototype.toString is not in the standard library, although I believe either LLVM or GNU has a non-standard extension, but I can't find it right now.*

**Comment by sffc** [comment]
Tentative decisions from meeting on 2020-07-17:

- Expose `impl From<f64/f32> for FixedDecimal` with a feature flag, calling Ryū under the hood.
- Don't implement the ICU4C fastpath algorithm.  Users who care about top speed and code size should architect their code to give us integers, not floats.

**Comment by sffc** [comment]
Additional thoughts:

1. Any API for this conversion should mandatorily take a number of decimal places, since trailing zeros are very important for i18n applications.
2. It's not really all that difficult to just punt this problem to user land; people can call Ryū directly (or choose their favorite algorithm) and then do string-to-FixedDecimal.

@filmil

**Comment by filmil** [comment]
> 2. It's not really all that difficult to just punt this problem to user land; people can call Ryū directly (or choose their favorite algorithm) and then do string-to-FixedDecimal.

I thought about this, and perhaps it is the best way forward. When writing an implementation of `select`, I could not use `format!` because the `format!` macro only supports static formatting.  But, the user always knows their format, so it would be trivial for them to use `format!` to convert a float into a representation they want formatted.

tl;dr: perhaps the best option for select() would be to take a formatted string instead of a float. (`FixedDecimal`) is not an option, as the trait definitions are supposed to not have significant dependencies.

**Comment by sffc** [comment]
The 402 trait accepting a decimal string SGTM.

**Comment by filmil** [comment]
> The 402 trait accepting a decimal string SGTM.

I see now that your original comment has a *footnote section written in cursive* which says essentially the same thing.  I missed that bit orignially.

**Comment by sffc** [comment]
@shadaj

**Comment by sffc** [comment]
@shadaj opened #724, but discovered that the `ryu` crate does not expose the API that we want.  In particular, we want an API that gives us the mantissa and exponent as integers, rather than a string.

We have a few options for our path forward.

### Option 1: Add API to `ryu`

We attempted to persue this option via discussions upstream in https://github.com/dtolnay/ryu/issues/27.  However, the maintainer of `ryu` has declined our requests to add the API.  This option can therefore be eliminated from consideration.

### Option 2: use `ryu_fixed_decimal`

The crate `ryu_fixed_decimal` is a hard fork of `ryu` that adds the functionality we need.

#### Option 2a: use `ryu_fixed_decimal` directly

The crate has one maintainer and not much activity.  It may be risky to add a dependency from ICU4X onto this type of crate.

#### Option 2b: adopt `ryu_fixed_decimal`

We could reach out to the maintainer of `ryu_fixed_decimal` and consider adopting it as an official ICU4X project.  This would come at an uncertain long-term maintenance cost.

### Option 3: use sub-optimal APIs of `ryu`

We could use `ryu`, but use the string-based API, meaning we would need to format the double to a string and then parse the string to get the digits.  @shadaj has found that this incurs a 36% performance hit, in addition to being just a generally inelegant solution.

### Option 4: use standard library

The standard library comes with its own double-to-string code.  We could use this instead of `ryu`.  This is what serde does.  It would still require writing the same string parsing code as option 3.  This is what the `bigdecimal` crate does ([link](https://docs.rs/bigdecimal/0.2.0/src/bigdecimal/lib.rs.html#1785-1796)).

It appears that the Rust standard library implements Grisu3, which is the same algorithm ICU4C uses.

There is additional internal unstable API in `std::num`:

https://ticki.github.io/redocs/std/num/flt2dec/index.html

### Option 5: use another Rust solution

The `rust_decimal` crate has its own algorithm that we could perhaps leverage:

https://docs.rs/rust_decimal/1.14.2/src/rust_decimal/decimal.rs.html#1112

There is dtolnay's other crate, `dtoa`:

https://docs.rs/dtoa/0.4.8/dtoa/

The `lexical` crate implements the Grisu2 algorithm:

https://docs.rs/lexical/5.2.2/lexical/index.html

### Option 6: Documentation

The easiest, and perhaps best, solution would be to basically say that this routine does not belong in ICU4X.  We can provide thorough examples of how to format doubles by leveraging your choice of double-to-decimal algorithm.

Option 6 could still involve adding functionality to parse the standard library or Ryu string format for ease of use.

**Comment by sffc** [comment]
2021-06-17:

- @Manishearth - Philosophically I like "bring your own crate". It affects FFI though. You need to own the FFI for those functions. But if we owned it, then we could build the FFI for it. With option 7, we basically say that you need to use your own language's double-to-decimal.
- @sffc - In JavaScript at least, that's a non-issue thanks to Number.toString().
- @zbraniecki - I like 7 for being Rusty, but I don't know how far we can go toward teaching the world not to use f64 as pluralization input. We could perhaps start with 7 and wait for customers to show up.
- @shadaj - The issue with converting inputs to strings is that you have to do additional parsing complexity like parsing the exponent field. So I think it's probably more than a 30% hit.
- @sffc - I think it's useful to add string parsing as a lowest common denominator.
- @Manishearth - Users shouldn't have to think about this much, because i18n is hard, and if we make them have to choose, they might not know enough to choose. But I like the idea of string API, plus, if you want something better, use one of these crates.

Conclusion:

- Adopt option 7, including the FromString impl for FixedDecimal
- Add good documentation aligning with the comments above

**Comment by sffc** [comment]
Note: We actually already have a FromString impl for FixedDecimal, but it doesn't support the exponent field.

I want to add the exponent field, but I've been thinking that we should preserve the exponent field in the data model of FixedDecimal, such that we can roundtrip "1.2e3" to "1.2e3".  We will be able to use this model to power compact decimal notation and scientific notation, and it is also useful as input to plural rules.

**Comment by sffc** [comment]
Discussion 2021-07-29:

- @nordzilla - Implementing this in SpiderMonkey, I think ICU4X should do more to handle this problem.
- @zbraniecki - I'm swayed.
- @echeran - Could we have a separate function for double-to-decimal instead of baking it in?
- @nordzilla - My understanding is that Option 3 means that ICU4X would provide a function to convert a double to a string with additional options like precision, trailing zeros, etc.
- @sffc - I think we're talking about double to FixedDecimal.
- @sffc - I think Option 3-6 will allow us to provide an API that helps avoid common pitfalls.
- @dminor - I think having this build in will make it more ergonomic for everyone.

Consensus: Proceed with Option 3.

**Comment by sffc** [comment]
For background / previous attempt, see #724

**Comment by Manishearth** [comment]
So I feel like we already have a "builder pattern" approach to `FixedDecimal`, and we're adding more such things in https://github.com/unicode-org/icu4x/pull/1195: I'd really prefer for these to be handled in a separate step, so for example you can call: `from_f32()`, and _follow that up_ with a call to `.round_digits()` or `truncate_digits()` (and then finally follow it up with padding calls if necessary). In fact this handles my complaint with https://github.com/unicode-org/icu4x/pull/1195 better than my suggestion there (to split up padding/truncate functions but have them work mostly the same).

Furthermore, enums like that are tricky to handle around FFI, so we'd need some sort of step-by-step API _anyway_. I don't mind adding an API that does `from_f32_auto_precision()` with separate `.round_to()` and `.truncate_to()` functions that we can combine into the fancy API, but I think the low level step by step API should exist and should be what we expose over FFI.

**Comment by sffc** [comment]
I understand where you're coming from.  But the issue is really twofold:

1. Theoretical purity: it is undefined to create a FixedDecimal from a Double without the precision, since a Double does not carry precision
2. Performance: we can, as ICU does, optimize the code path when we have the double and the precision in the same function

**Comment by sffc** [comment]
If we wanted a true "builder" type pattern, then we could consider an intermediate class called something like `FixedDecimalFromDoubleBuilder` that resolves to a `FixedDecimal` once the precision is specified.  This is actually what ICU does under the hood, although it does so with a boolean flag on the fixed decimal type instead of a separate type.

**Comment by zbraniecki** [comment]
I'm with Shane here. I am not a huge fan of builder patterns in general, but when used, I really prefer to never operate on half baked states. If the only thing you can do with a state is to add precision to it to produce fixed decimal, then this intermediate step is pointless and harmful from the API design perspective.

So, even if you do go for a builder pattern here, I'd suggest that we either allow for precisionless construction of FixedDecimal (with known limitations of that approach and strong recommendation to use a precision-full constructor), or only allow for construction with value+precision provided at the same time.

**Comment by Manishearth** [comment]
So I'll respond to some of the points later when I get the chance, but a thing I want to point out quickly is that while floating point numbers have no inherent decimal precision, ryū's model for them _does_, so we may need to reevaluate using ryū here if we want this level of control. At the moment the arguments for such an API do not make sense _in the context of our double to decimal library being ryū_, which is why I prefer the builder design. If we want something else, we should pick a different double-to-decimal routine; maybe write one ourselves.

**Comment by sffc** [comment]
> while floating point numbers have no inherent decimal precision, ryū's model for them _does_

ryū, as well as other algorithms like grisu3, implements an algorithm that returns a decimal string with the fewest number of digits required to unambiguously recover the original decimal.  This is different than the concept of "precision" that I'm discussing.

One valid precision value is "maximum supported by the floating point type I am using".  This is what a vanilla ryū-based conversion performs.  But this is not the only precision strategy.  I also don't want to give this precision strategy favoritism over other precision strategies.

**Comment by Manishearth** [comment]
@sffc right, I'm saying that ryū is not sufficient for us to support these other options since it has a different strategy here. We need something more, and I'm asking where we get that from, because otherwise this can't be implemented.

I'm trying to find an MVP that supports WearOS' use case and also find what we need to pull in to support the full API here.

MVP wise I think we may be able to use ryū as the Maximum mode with a separate function for truncation and rounding, and we can add a full version later that the Maximum mode function calls.

**Comment by Manishearth** [comment]
Update from the discussion I had with @sffc:

I think what I'd misunderstood here was an implication that ryū could not be used as a base for the rounding/truncating cases because ryū can "shorten" some floats. It seems like that's not central to this; so we should be able to implement Shane's proposed API, with less enumy versions for FFI.

**Comment by Manishearth** [comment]
A hitch: ryū formats scientific notation in its double-to-decimal algorithm, we may need `from_str` to handle that as well

**Comment by shadaj** [comment]
> A hitch: ryū formats scientific notation in its double-to-decimal algorithm, we may need `from_str` to handle that as well

Yeah, this is why I dropped using the main ryū crate in favor of the fork that just gives us an integer when putting together the PR for this issue.

**Comment by Manishearth** [comment]
Opened https://github.com/unicode-org/icu4x/pull/1265 to handle exponents
